api/index.php

```diff
@@ -0,0 +1,203 @@
+{
+    $username = api_authenticate();
+
+    api_rate_limit($username);
+
+    $query_params = $_GET;
+    $path = explode('/', array_get($query_params, "url", ""));
+    unset($query_params["url"]);
+
+    // if the path ended with a /, pop it off for correct routing
+    if($path[count($path) - 1] == '')
+    {
+        array_pop($path);
+    }
+
+    api_output_response(api_router($path, $query_params));
+}
+
+function api_authenticate()
+{
+    global $pguser;
+
+    $api_key = @$_SERVER['HTTP_X_API_KEY'];
+    if($api_key)
+    {
+        try {
+            $user = User::load_from_api_key($api_key);
+        } catch (NonexistentUserException $exception) {
+            throw new UnauthorizedError();
+        }
+        $pguser = $user->username;
+    }
+
+    if(!isset($pguser))
+    {
+        throw new UnauthorizedError();
+    }
+
+    return $pguser;
+}
+
+function api_rate_limit($key)
+{
+    global $api_rate_limit;
+    global $api_rate_limit_requests_per_window, $api_rate_limit_seconds_in_window;
+
+    if(!$api_rate_limit)
+    {
+        return;
+    }
+
+    $memcache = new Memcached();
+    $memcache->addServer('localhost', 11211);
+
+    // increment (or initialize) our count
+    $count = $memcache->get("$key:count");
+    $count = $count === FALSE ? 0 : $count + 1;
+    if($memcache->set("$key:count", $count) === FALSE)
+    {
+        // if we can't set the value, memcached probably isn't running
+        // regardless, we can't enforce rate limiting so return an error
+        throw new ServerError();
+    }
+
+    // initialize or reset our expire time
+    $expire_time = $memcache->get("$key:expire");
+    if($count === 0 || $expire_time < time())
+    {
+        $expire_time = time() + $api_rate_limit_seconds_in_window;
+        $memcache->set("$key:expire", $expire_time);
+    }
+
+    // see if we need to reset our count
+    $seconds_before_reset = $expire_time - time();
+    if($seconds_before_reset <= 0)
+    {
+        $memcache->set("$key:count", 0);
+    }
+
+    // enforce exceeding the limit
+    if($count > $api_rate_limit_requests_per_window)
+    {
+        throw new RateLimitExceeded(
+            "Rate limit exceeded, resets in $seconds_before_reset seconds"
+        );
+    }
+
+    $requests_left_before_reset = $api_rate_limit_requests_per_window - $count;
+    header("X-Rate-Limit-Limit: $api_rate_limit_requests_per_window");
+    header("X-Rate-Limit-Remaining: $requests_left_before_reset");
+    header("X-Rate-Limit-Reset: $seconds_before_reset");
+}
+
+function api_router($path, $query_params, ...$data)
+{
+    $trace = debug_backtrace();
+    $last_hop = $trace[1]["function"];
+
+    $next_hop = array_shift($path);
+    $next_hop_function = "${last_hop}_${next_hop}";
+    if(!$next_hop || !function_exists($next_hop_function))
+    {
+        $path_url = str_replace("_", "/", $next_hop_function);
+        throw new InvalidAPI();
+    }
+
+    return $next_hop_function($path, $query_params, $data);
```

Comment:
This aspect of PHP terrifies me, but hey, it is what it is. At least we have api_validate_method below to stop API clients from calling arbitrary functions...

Reply:
I am very open to other approaches to an implementation abstraction that doesn't require every endpoint to validate arguments. For instance:
```
v1/projects/:projectid
v1/projects/:projectid/wordlists/good
```
^^^ I don't want to have both of those do independent validations of the projectid.

Reply:
The new implementation is much better and doesn't rely on dynamic functions.